internal/gitprovider/github/github.go

```diff
@@ -0,0 +1,164 @@
+package github
+
+import (
+	"context"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/google/go-github/v56/github"
+	"github.com/pkg/errors"
+	"k8s.io/utils/pointer"
+
+	"github.com/akuity/kargo/internal/gitprovider"
+)
+
+var (
+	githubRegistration = gitprovider.ProviderRegistration{
+		Predicate: func(repoURL string) bool {
+			u, err := url.Parse(repoURL)
+			if err != nil {
+				return false
+			}
+			// The assumption here is that no one would ever put "github" in
+			// the DNS name of their self-hosted git provider. There is also
+			// the chance that they self host github but give it a name like:
+			// 'git.mycompany.com', in which case we would return false
+			return strings.Contains(u.Host, "github")
+		},
+		NewService: func() (gitprovider.GitProviderService, error) {
+			return NewGitHubProvider()
+		},
+	}
+)
+
+func init() {
+	gitprovider.RegisterProvider("github", githubRegistration)
+}
+
+type GitHubProvider struct {
+	client *github.Client
+}
+
+func NewGitHubProvider() (*GitHubProvider, error) {
```

Comment:
I think it would be an improvement if this function signature returned a `gitprovider.GitProviderService` interface instead of directly returning a specific implementation.

By doing so, the implementation can also be made unexported and implementation details, therefore, neatly hidden from callers.